pkg/apis/pingcap/v1alpha1/tidbcluster_strategy.go

```diff
@@ -0,0 +1,205 @@
+// Copyright 2019. PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"context"
+	"reflect"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/klog"
+)
+
+const (
+	defaultTiDBImage   = "pingcap/tidb"
+	defaultTiKVImage   = "pingcap/tikv"
+	defaultPDImage     = "pingcap/pd"
+	defaultBinlogImage = "pingcap/tidb-binlog"
+)
+
+// +k8s:deepcopy-gen=false
+type TidbClusterStrategy struct{}
+
+func (TidbClusterStrategy) NewObject() runtime.Object {
+	return &TidbCluster{}
+}
+
+func (TidbClusterStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
+	if tc, ok := castTidbCluster(obj); ok {
+		setTidbClusterDefault(tc)
+	}
+}
+
+func (TidbClusterStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
+	// no op to not affect the cluster managed by old versions of the helm chart
+}
+
+func (TidbClusterStrategy) Validate(ctx context.Context, obj runtime.Object) field.ErrorList {
+	if tc, ok := castTidbCluster(obj); ok {
+		return validateTidbCluster(tc)
+	}
+	return field.ErrorList{}
+}
+
+func (TidbClusterStrategy) ValidateUpdate(ctx context.Context, obj, old runtime.Object) field.ErrorList {
+	oldTc, oldOk := castTidbCluster(old)
+	tc, ok := castTidbCluster(obj)
+	if ok && oldOk {
+		return validateUpdateTidbCluster(oldTc, tc)
+	}
+	return field.ErrorList{}
+}
+
+func setTidbClusterDefault(tc *TidbCluster) {
+	if tc.Spec.TiDB.BaseImage == "" {
+		tc.Spec.TiDB.BaseImage = defaultTiDBImage
+	}
+	if tc.Spec.TiKV.BaseImage == "" {
+		tc.Spec.TiKV.BaseImage = defaultTiKVImage
+	}
+	if tc.Spec.PD.BaseImage == "" {
+		tc.Spec.PD.BaseImage = defaultPDImage
+	}
+	if tc.Spec.Pump != nil && tc.Spec.Pump.BaseImage == "" {
+		tc.Spec.Pump.BaseImage = defaultBinlogImage
+	}
+	if tc.Spec.TiDB.Config == nil {
+		tc.Spec.TiDB.Config = &TiDBConfig{}
+	}
+	if tc.Spec.TiKV.Config == nil {
+		tc.Spec.TiKV.Config = &TiKVConfig{}
+	}
+	if tc.Spec.PD.Config == nil {
+		tc.Spec.PD.Config = &PDConfig{}
+	}
+	if string(tc.Spec.ImagePullPolicy) == "" {
+		tc.Spec.ImagePullPolicy = corev1.PullIfNotPresent
+	}
+}
+
+func validateTidbCluster(tc *TidbCluster) field.ErrorList {
+	allErrs := validateTidbClusterSpec(&tc.Spec, field.NewPath("spec"))
+	return allErrs
+}
```

Comment:
put validation logic into validation pkg: https://github.com/pingcap/tidb-operator/blob/master/pkg/apis/pingcap/v1alpha1/validation? easier to test when it got complicated

Reply:
Good advice!

Reply:
Addressed in https://github.com/pingcap/tidb-operator/pull/1429/commits/60d3c5bf860ebba5a8877b529d633439a199a28d